Added solution to Word Ladder

amemov opened the pull request.

amemov committed:
Added solution to Word Ladder

amemov committed:
- Replaced <unordered_map> with absl::flat_hash_map
- Replaced int with size_t for assign op with .size()
- Refactroed the code: removed unneceray class declaration
- Reformatted variables and function names in snake_case to match Google Format
- Removed C-Style comments /**/
- Modified comments in definition and implementation of API: header says what function computes and which leetcode problem it solves, cpp explains in detail the approach to the problem

amemov commented:
Reviewed the code: addressed some small problems with types, refactored the structure, and reformatted the names to match Google's Format

kcc commented on Word Ladder/BUILD.bazel:
```diff
@@ -0,0 +1,20 @@
+
+cc_test(
+    name = "solution_test",
+    size = "small",
```
redundant - drop

kcc commented on Word Ladder/MODULE.bazel.lock:
you probably don't want this file in the PR

kcc commented on Word Ladder/readme.md:
````diff
@@ -0,0 +1,6 @@
+
+To run, type the following in the terminal:
+```
+bazel test --cxxopt=-std=c++17 --test_output=all //:solution_test
````
Note: if you want to run a single test and see the output, you can use `bazel run`

kcc commented on Word Ladder/solution.h:
```diff
@@ -0,0 +1,17 @@
+// This function return the number of words in the shortest transformation 
+// sequence from begin_word to end_word. Returns 0 if no such sequence.
+// Word Ladder Problem (LeetCode 127)
+int LadderLength(std::string begin_word, std::string end_word,
```
prefer to pass arguments as `const std::string &` or (better) `std::string_view`.
The comment does not explain the meaning of `word_list`, 
It it's a dictionary, it should be const.

kcc commented on Word Ladder/solution.cpp:
```diff
@@ -0,0 +1,74 @@
+  // Create hashMap to store possible paths for every letter change
+  // for each word
+  absl::flat_hash_map<std::string, std::vector<std::string> > neighbors_map;
+  int word_len = begin_word.length();  // words are of equal length
```
if anything is a size/length, its type is size_t

kcc replied:
use CHECK_EQ to verify that the two words are of equal length. 
Or return 0 if not.

amemov replied:
Problem constraints gurantee us that the begin_word and end_word are of equal length. Same applies to words inside word_list vector - all words are guranteed to be of equal length, according to the problem definition on the LeetCode. In this case I don't think checking lengths would be necessary, nonetheless, it is a reasonable edge case.

kcc commented on Word Ladder/solution.cpp:
```diff
@@ -0,0 +1,74 @@
+  absl::flat_hash_map<std::string, std::vector<std::string> > neighbors_map;
+  int word_len = begin_word.length();  // words are of equal length
+
+  // Go through each word and create store possible patterns
```
typo?

kcc commented on Word Ladder/solution.cpp:
```diff
@@ -0,0 +1,74 @@
+  // Go through each word and create store possible patterns
+  // associated with a given word
+  // E.g neighbors_map[h*t] = {hot, hit}
+  for (std::string word_ : word_list) {
```
don't iterate using std::string: it creates a copy. 
use `for (auto & word: word_list)`

Also, check the variable naming

kcc commented on Word Ladder/solution.cpp:
```diff
@@ -0,0 +1,74 @@
+  // associated with a given word
+  // E.g neighbors_map[h*t] = {hot, hit}
+  for (std::string word_ : word_list) {
+    for (int i = 0; i < word_len; ++i) {
```
size_t

kcc commented on Word Ladder/solution.cpp:
```diff
@@ -0,0 +1,74 @@
+      // Given combinations, see where we can go and only go through elements
+      // That we haven't added to visited_set yet
+      for (int j = 0; j < word_len; ++j) {
+        std::string pattern_ =
```
this pattern appears twice - move it to a separate function or a local lambda

kcc commented on Word Ladder/solution_test.cpp:
```diff
@@ -0,0 +1,44 @@
+  EXPECT_EQ(LadderLength(begin_word, end_word, word_list), 0);
+}
+
+// Main function to run all tests
```
you have         "@googletest//:gtest_main", in deps, you don't need main() here

kcc commented on Word Ladder/solution.cpp:
```diff
@@ -0,0 +1,74 @@
+  // E.g neighbors_map[h*t] = {hot, hit}
+  for (std::string word_ : word_list) {
+    for (int i = 0; i < word_len; ++i) {
+      std::string pattern_ =
```
I would write this code like this: 
   std::string pattern = word;
   pattern[i] = '*'

kcc commented on Word Ladder/solution.cpp:
```diff
@@ -0,0 +1,74 @@
+  // If can't, while-loop will eventually run out of element in words_queue 
+  // and return 0
+  int shortest_path = 1;
+  std::set<std::string> visited_set;
```
you can also use flat_hash_set.
It's faster and has .contains()

amemov replied:
That's a good point about .contains().
I initially tried it with .find() and it showed 3 times slow-down over std::set, so I switched back to std::set. Now I tested flat_hash_set with .contains() and it works better

kcc commented on Word Ladder/solution.cpp:
```diff
@@ -0,0 +1,74 @@
+        std::string pattern_ =
+            current_word.substr(0, j) + '*' + current_word.substr(j + 1, word_len);
+        
+        for (std::string neighbor_word : neighbors_map[pattern_]) {
```
for (const auto & var :

kcc reviewed (commented):
Yep, looks like a valid solution. Added a few style comments.

amemov committed:
Deleted .bazel.lock file from PR

amemov committed:
Addressed the comments:
- Added more comments to explain the meaning of word_list
- Removed redundancy and unnecessary files
- Fixed some variables names
- Fixed for-loop iterations, removing unnecessary string creations
- Added additional base case check
- Replaced 'int' with 'size_t' where needed
- Wrote a local lamda function for pattern generation
- Changed argument definition in LadderLength
- Replaced std::set with absl::flat_hash_set

amemov committed:
Merge branch 'wordLadder' of https://github.com/amemov/Coding-Puzzles-Solutions into wordLadder

amemov committed:
- Changed #include order for solution.h

kcc commented on Word Ladder/BUILD.bazel:
```diff
@@ -4,13 +4,12 @@ cc_library(
     deps = [ 
         "@abseil-cpp//absl/container:flat_hash_map",
+        "@abseil-cpp//absl/container:flat_hash_set",
+        "@com_github_google_glog//:glog",
```
glog is deprecated, use absl/log/log.g

kcc commented on Word Ladder/solution.cpp:
```diff
@@ -0,0 +1,83 @@
+  int shortest_path = 1;
+  absl::flat_hash_set<std::string> visited_set;
+  std::deque<std::string> words_queue;
+  words_queue.push_back(static_cast<std::string>(begin_word));
```
don't cast from string_view, create from it: std::string(begin_word)

kcc reviewed (commented):
this is pretty good. If you want to do another problem like this, please try to not look at other existing solution at all, and come up with your own instead.

amemov committed:
Addressed the final comments:
- Replaced GLOG with Abseil. Updated dependencies caused by this change
- Replaced static_cast with std::string
